### PR TITLE
Group Norm

### PR DESCRIPTION
**Pull request by cjld**
This pr is still WIP. currently, only the CPU version is supported.

### CODE REVIEW

**Comment by cjld** [comment]
Though this PR does not reach peak performance(60% currently). But it is ready for review.

File: paddle/fluid/operators/group_norm_op.cc [hunk 1]
```diff
@@ -0,0 +1,163 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline paddle/fluid/operators/group_norm_op.cc]: 2016 ->  2018

File: paddle/fluid/operators/group_norm_op.cc [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+    AddAttr<float>("epsilon",
+                   "Constant for numerical stability [default 1e-5].")
+        .SetDefault(1e-5)
+        .AddCustomChecker([](const float &epsilon) {
+          PADDLE_ENFORCE(epsilon >= 0.0f && epsilon <= 1.0f,
+                         "'epsilon' should be between 0.0 and 1.0.");
+        });
+    AddAttr<int>("groups", "The number of groups that divided from channels.")
+        .SetDefault(32)
```
Comment (qingqing01) [inline paddle/fluid/operators/group_norm_op.cc]: Maybe not set defalut value.

File: paddle/fluid/operators/group_norm_op.cu [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+    // init output
+    auto* d_x = ctx.Output<Tensor>(framework::GradVarName("X"));
+    auto* d_scale = ctx.Output<Tensor>(framework::GradVarName("Scale"));
+    auto* d_bias = ctx.Output<Tensor>(framework::GradVarName("Bias"));
+
+    const auto& x_dims = x->dims();
+    const int group_size = (x_dims[1] - 1) / groups + 1;
+
+    // TODO(liangdun): need to check d_x is null
```
Comment (qingqing01) [inline paddle/fluid/operators/group_norm_op.cu]: need to fix this.

File: paddle/fluid/operators/group_norm_op.h [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        T x_mean = 0, x_var = 0;
+        int number = std::min(group_size,
+                              static_cast<int>(x_dims[1] - gid * group_size));
+        auto* tmp = iter_x_data;
+        for (int cid = 0; cid < number; cid++) {
+          for (int imid = 0; imid < imsize; imid++, iter_x_data++) {
+            x_mean += iter_x_data[0];
+            x_var += iter_x_data[0] * iter_x_data[0];
+          }
+        }
```
Comment (qingqing01) [inline paddle/fluid/operators/group_norm_op.h:75]: For cpu, can used Eigen for some vector calculation.
or some fucntion in paddle/fluid/operators/math/blas.h

File: paddle/fluid/operators/group_norm_op.h [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+    // init output
+    auto* d_x = ctx.Output<Tensor>(framework::GradVarName("X"));
+    auto* d_scale = ctx.Output<Tensor>(framework::GradVarName("Scale"));
+    auto* d_bias = ctx.Output<Tensor>(framework::GradVarName("Bias"));
+
+    const auto& x_dims = x->dims();
+    const int group_size = (x_dims[1] - 1) / groups + 1;
+
+    // TODO(liangdun): need to check d_x is null
```
Comment (qingqing01) [inline paddle/fluid/operators/group_norm_op.h:114]: need to fix this.

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -2305,6 +2306,87 @@ def layer_norm(input,
...
+               name=None):
+    """
+    **Group Normalization Layer**
+
+    Refer to `Group Normalization <https://arxiv.org/abs/1803.08494>`
+
+    Args:
+        input(Variable): The input tensor variable.
+        scale(bool): Scale is a 1-dimensional tensor of size C that is multiply to the output.
+        shift(bool): Shift is a 1-dimensional tensor of size C that is add to the output.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: There is no need to use `scale(bool)` and ` shift(bool)`, can use `param_attr` and `bias_attr` to determine whether to apply scale and bias.  The `param_attr` and `bias_attr` also can be set True or False in other APIs, like fc.

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -2305,6 +2306,87 @@ def layer_norm(input,
...
+        groups(int): The number of groups that divided from channels.
+        epsilon(float): The small value added to the variance to prevent
+            division by zero.
+        param_attr(ParamAttr|None): The parameter attribute for the learnable
+            gain :math:`g`.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: gain -> scale?

File: python/paddle/fluid/layers/nn.py [hunk 8]
```diff
@@ -2305,6 +2306,87 @@ def layer_norm(input,
...
+        bias_attr(ParamAttr|None): The parameter attribute for the learnable
+            bias :math:`b`.
+        act(str): Activation to be applied to the output of group normalizaiton.
+        data_layout(string|NCHW): Only NCHW is supported.
+        name (str): The name of this layer. It is optional.
+
+    Returns:
+        ${y_comment}
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Give the commentis, instead ${y_comment}, keep consistent with the above variables.

File: python/paddle/fluid/layers/nn.py [hunk 9]
```diff
@@ -2305,6 +2306,87 @@ def layer_norm(input,
     return helper.append_activation(layer_norm_out)
 
 
+@templatedoc()
+def group_norm(input,
+               scale=True,
+               shift=True,
+               groups=32,
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: maybe not set defalut number for `groups` ?

**Comment by qingqing01** [comment]
@cjld Please update your code.

File: paddle/fluid/operators/group_norm_op.h (hunk 4 above)
Comment (cjld) [reply paddle/fluid/operators/group_norm_op.h:75]: I am not very familiar with Eigen API. This issue may be fixed in next PR.

File: python/paddle/fluid/layers/nn.py [hunk 10]
```diff
@@ -2305,6 +2306,82 @@ def layer_norm(input,
...
+
+    Args:
+        input(Variable): The input tensor variable.
+        groups(int): The number of groups that divided from channels.
+        epsilon(float): The small value added to the variance to prevent
+            division by zero.
+        param_attr(ParamAttr|None): The parameter attribute for the learnable
+            scale :math:`g`.
+        bias_attr(ParamAttr|None): The parameter attribute for the learnable
+            bias :math:`b`.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Please refer bias_attr comment in `fc`, need detailed info.

**Review by qingqing01 — DISMISSED** [review]

File: python/paddle/fluid/layers/nn.py (hunk 10 above)
Comment (cjld) [reply python/paddle/fluid/layers/nn.py]: fixed

**Review by qingqing01 — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]
LG API

**Pull request merged**